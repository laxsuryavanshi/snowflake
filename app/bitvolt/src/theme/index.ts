export * from './hooks';
export { default as ThemeContext } from './ThemeContext';
export * from './ThemeContext';
export { default as ThemeProvider } from './ThemeProvider';
