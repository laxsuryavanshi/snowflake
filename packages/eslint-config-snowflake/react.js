import react from 'eslint-plugin-react';
import reactHooks from 'eslint-plugin-react-hooks';
import reactRefresh from 'eslint-plugin-react-refresh';

import config from './index.js';

export default [
  {
    plugins: {
      react,
      'react-hooks': reactHooks,
      'react-refresh': reactRefresh,
    },
    rules: {
      ...reactHooks.configs.recommended.rules,
      'react-refresh/only-export-components': ['warn', { allowConstantExport: true }],
    },
  },
  ...config,
];
