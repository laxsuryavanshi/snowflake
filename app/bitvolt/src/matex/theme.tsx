import { createContext, useContext, useMemo } from 'react';
import type { PaletteMode, Theme } from '@mui/material';
import { useMediaQuery } from '@mui/material';
import CssBaseline from '@mui/material/CssBaseline';
import type { LinkProps } from '@mui/material/Link';
import {
  createTheme as createMuiTheme,
  ThemeProvider as MuiThemeProvider,
  responsiveFontSizes,
} from '@mui/material/styles';
import { useLocalStorage } from 'usehooks-ts';

import Link from './Link';

export type DisplayMode = PaletteMode | 'system';

export interface ThemeProviderValue {
  get mode(): DisplayMode;

  setMode: (mode: DisplayMode) => void;
}

const ThemeContext = createContext<ThemeProviderValue | undefined>(undefined);

const createTheme = (mode: PaletteMode): Theme => {
  let theme = createMuiTheme({
    typography: {
      fontFamily: 'var(--snowflake-font-family)',
      fontWeightMedium: 600,
      button: {
        textTransform: 'none',
        fontWeight: 'inherit',
      },
    },
    palette: { mode },
    components: {
      MuiLink: {
        defaultProps: {
          component: Link,
        } as LinkProps,
      },
      MuiButtonBase: {
        defaultProps: {
          LinkComponent: Link,
        },
      },
    },
  });

  theme = responsiveFontSizes(theme);

  return theme;
};

export const useSystemDisplayMode = (): PaletteMode => {
  return useMediaQuery('(prefers-color-scheme: dark)') ? 'dark' : 'light';
};

export const ThemeProvider: React.FC<React.PropsWithChildren> = ({ children }) => {
  const systemDisplayMode = useSystemDisplayMode();
  const [displayMode, setDisplayMode] = useLocalStorage<DisplayMode>('displayMode', 'system');
  const [displayModeInternal, setDisplayModeInternal] = useLocalStorage<PaletteMode>(
    'displayModeInternal',
    systemDisplayMode
  );

  const setMode = (mode: DisplayMode) => {
    setDisplayMode(mode);

    if (mode === 'system') {
      setDisplayModeInternal(systemDisplayMode);
    } else {
      setDisplayModeInternal(mode);
    }
  };

  const theme = useMemo(() => createTheme(displayModeInternal), [displayModeInternal]);
  const value = useMemo(() => ({ mode: displayMode, setMode }), [displayMode]);

  return (
    <ThemeContext.Provider value={value}>
      <MuiThemeProvider theme={theme}>
        <CssBaseline />
        {children}
      </MuiThemeProvider>
    </ThemeContext.Provider>
  );
};

export const useTheme = (): ThemeProviderValue => {
  const context = useContext(ThemeContext);

  if (!context) {
    throw new Error('`useTheme` must be used inside a `ThemeProvider`');
  }

  return context;
};
