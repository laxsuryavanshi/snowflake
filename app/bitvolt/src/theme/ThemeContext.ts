import { createContext } from 'react';
import type { PaletteMode } from '@mui/material';

export type DisplayMode = PaletteMode | 'system';

export interface ThemeProviderValue {
  get mode(): DisplayMode;

  setMode: (mode: DisplayMode) => void;
}

const ThemeContext = createContext<ThemeProviderValue | undefined>(undefined);

export default ThemeContext;
