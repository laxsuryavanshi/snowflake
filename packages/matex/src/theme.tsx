import { createContext, useContext, useMemo } from 'react';
import { PaletteMode, Theme, useMediaQuery } from '@mui/material';
import CssBaseline from '@mui/material/CssBaseline';
import { LinkProps } from '@mui/material/Link';
import {
  ThemeProvider as MuiThemeProvider,
  createTheme as createMuiTheme,
  responsiveFontSizes,
} from '@mui/material/styles';
import { useLocalStorage } from 'usehooks-ts';

import Link from './Link';

export type DisplayMode = PaletteMode | 'system';

export interface IThemeContext {
  get mode(): DisplayMode;

  setMode: (mode: DisplayMode) => void;
}

/**
 * Creates a Material-UI theme object based on the specified mode (light or dark).
 * @param {PaletteMode} mode The mode of the theme (light or dark).
 * @returns {Theme} The Material-UI theme object.
 */
const createTheme = (mode: PaletteMode): Theme => {
  let theme = createMuiTheme({
    typography: {
      fontFamily: 'var(--snowflake-font-family)',
      fontWeightBold: 700,
      button: {
        textTransform: 'none',
        fontWeight: 'inherit',
      },
    },
    palette: { mode },
    components: {
      MuiLink: {
        defaultProps: {
          component: Link,
        } as LinkProps,
      },
      MuiButtonBase: {
        defaultProps: {
          LinkComponent: Link,
        },
      },
    },
  });

  theme = responsiveFontSizes(theme);

  return theme;
};

const ThemeContext = createContext<IThemeContext | null>(null);

/**
 * Returns the current system display mode (light or dark) based on the user's preferences.
 * @returns {PeletteMode} The current system display mode.
 */
const useSystemDisplayMode = (): PaletteMode =>
  useMediaQuery('(prefers-color-scheme: dark)') ? 'dark' : 'light';

export const ThemeProvider = ({ children }: React.PropsWithChildren) => {
  const systemDisplayMode = useSystemDisplayMode();

  /**
   * @var displayMode is used to determine the mode user has selected in our application
   */
  const [displayMode, setDisplayMode] = useLocalStorage<DisplayMode>('displayMode', 'system');
  /**
   * @var displayModeInternal unlike @var displayMode, is the actual mode applied in the
   * application. When displayMode is set to `system`, we use users' system preferences
   * to determine the settings to use in our app.
   */
  const [displayModeInternal, setDisplayModeInternal] = useLocalStorage<PaletteMode>(
    'displayModelInternal',
    systemDisplayMode
  );

  const setSystemDisplayMode = (mode: DisplayMode) => {
    if (mode === 'system') {
      mode = systemDisplayMode;
    }
    setDisplayModeInternal(mode);
  };
  const setMode = (mode: DisplayMode) => {
    setDisplayMode(mode);
    setSystemDisplayMode(mode);
  };

  const theme = useMemo(() => createTheme(displayModeInternal), [displayModeInternal]);
  // eslint-disable-next-line react-hooks/exhaustive-deps
  const value = useMemo(() => ({ mode: displayMode, setMode }), [displayMode]);

  return (
    <ThemeContext.Provider value={value}>
      <MuiThemeProvider theme={theme}>
        <CssBaseline />
        {children}
      </MuiThemeProvider>
    </ThemeContext.Provider>
  );
};

export const useTheme = (): IThemeContext => {
  const context = useContext(ThemeContext);

  if (!context) {
    throw new Error('`useTheme` must be used inside a `ThemeProvider`');
  }

  return context;
};
