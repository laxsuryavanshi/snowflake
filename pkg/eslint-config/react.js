import react from 'eslint-plugin-react';
import reactHooks from 'eslint-plugin-react-hooks';
import reactRefresh from 'eslint-plugin-react-refresh';

export default [
  {
    name: '@snowflake/react',
    plugins: {
      react,
      'react-hooks': reactHooks,
      'react-refresh': reactRefresh,
    },
    rules: {
      ...reactHooks.configs.recommended.rules,
      'react-refresh/only-export-components': ['warn', { allowConstantExport: true }],
    },
  },
];
