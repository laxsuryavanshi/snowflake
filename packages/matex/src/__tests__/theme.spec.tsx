import { act } from 'react';
import { renderHook } from '@testing-library/react';

import { ThemeProvider, useTheme } from '../theme';

const Wrapper = ({ children }: React.PropsWithChildren) => (
  <ThemeProvider>{children}</ThemeProvider>
);

describe('useTheme', () => {
  it('should throw an error when used outside a ThemeProvider', () => {
    expect(() => renderHook(() => useTheme())).toThrow(
      '`useTheme` must be used inside a `ThemeProvider`'
    );
  });

  it('should return mode and setter function', () => {
    const { result } = renderHook(() => useTheme(), { wrapper: Wrapper });

    expect(result.current).toMatchSnapshot();
  });

  it('should set display mode to dark', () => {
    const { result } = renderHook(() => useTheme(), { wrapper: Wrapper });

    expect(result.current.mode).toBe('system');
    act(() => {
      result.current.setMode('dark');
    });
    expect(result.current.mode).toBe('dark');
    act(() => {
      result.current.setMode('system');
    });
    expect(result.current.mode).toBe('system');
  });
});
