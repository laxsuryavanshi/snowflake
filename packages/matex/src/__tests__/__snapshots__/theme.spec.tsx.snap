// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useTheme should return mode and setter function 1`] = `
{
  "mode": "system",
  "setMode": [Function],
}
`;
